text/0000-unsafe-fields.md

```diff
@@ -0,0 +1,1037 @@
+    }
+
+    /// Produces a `KeepAlive` handle, which defers the destruction
+    /// of the enclosed value.
+    pub fn keep_alive(&self) -> KeepAlive<T> {
+        // SAFETY: By invariant on `KeepAlive::arc`, this clone will never be
+        // used for accessing `T`, as required by `UniqueArc::arc`. The one
+        // exception is that, if a `KeepAlive` is the last reference to be
+        // dropped, then it will drop the inner `T`. However, if this happens,
+        // it means that the `UniqueArc` has already been dropped, and so its
+        // invariant will not be violated.
+        unsafe {
+            KeepAlive {
+                arc: self.arc.clone(),
+            }
+        }
+    }
+}
+
+impl<T> Deref for UniqueArc<T> {
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        // SAFETY: We do not create any other owning references to `arc` - we
+        // only dereference it below, but do not clone it.
+        let arc = unsafe { &self.arc };
+        let ptr = UnsafeCell::get(arc);
+        // SAFETY: We satisfy all requirements for pointer-to-reference
+        // conversions [1]:
+        // - By invariant on `&UnsafeCell<T>`, `ptr` is well-aligned, non-null,
+        //   dereferenceable, and points to a valid `T`.
+        // - By invariant on `Self::arc`, no other `Arc` references exist to
+        //   this value which will be used for reading or writing. Thus, we
+        //   satisfy the aliasing invariant of `&` references.
+        //
+        // [1] https://doc.rust-lang.org/1.85.0/std/ptr/index.html#pointer-to-reference-conversion
+        unsafe { &*ptr }
+    }
+}
+
+impl<T> DerefMut for UniqueArc<T> {
+    fn deref_mut(&mut self) -> &mut T {
+        // SAFETY: We do not create any other owning references to `arc` - we
+        // only dereference it below, but do not clone it.
+        let arc = unsafe { &mut self.arc };
+        let val = UnsafeCell::get(arc);
+        // SAFETY: We satisfy all requirements for pointer-to-reference
+        // conversions [1]:
+        // - By invariant on `&mut UnsafeCell<T>`, `ptr` is well-aligned,
+        //   non-null, dereferenceable, and points to a valid `T`.
+        // - By invariant on `Self::arc`, no other `Arc` references exist to
+        //   this value which will be used for reading or writing. Thus, we
+        //   satisfy the aliasing invariant of `&mut` references.
+        //
+        // [1] https://doc.rust-lang.org/1.85.0/std/ptr/index.html#pointer-to-reference-conversion
+        unsafe { &mut *val }
+    }
+}
+```
+
+# Reference-level explanation
+
+## Syntax
+
+The [`StructField` syntax][struct syntax], used for the named fields of structs, enums, and unions,
+shall be updated to accommodate an optional `unsafe` keyword just before the field `IDENTIFIER`:
+
+```diff
+StructField :
+   OuterAttribute*
+   Visibility?
++  unsafe?
+   IDENTIFIER : Type
+```
+
+[struct syntax]: https://doc.rust-lang.org/stable/reference/items/structs.html#structs
```

Comment:
how about `TupleField`? (oversight? future extension? by-design unsupported because of ambiguity with `struct F(unsafe fn());`?)

Reply:
By-design unsupported because of ambiguity. Might be possible to disambiguate with parens, but that's future work.